### PR TITLE
feat: Add DiscountConditions

### PR DESCRIPTION
**Pull request by olivermrbl**
**What**
Introduces DiscountConditions and removes`valid_for`

**Why**
To enable more advanced promotions such as creating a discount for a specific set of products, collections, types, tags, and customer groups.

**How**
Introduce new entity DiscountCondition that holds a type ("products"), operator ("in" meaning resources must be in the condition for the discount to be valid), and a reference to a specific type of DiscountCondition e.g. DiscountConditionProduct. Read more about the entity implementation [here](https://github.com/medusajs/medusa/discussions/1148). #1146 

Convert DiscountService to TypeScript. #1149 

Include new conditions in payload to update and create discount. #1170 

Add DiscountCondition repository that allows us to add, update and remove conditions in bulk, and check the applicableness of a given line item or customer. #1156, #1172, #1212 

Refactor the discount calculation for a LineItem from TotalsService to DiscountService. #1224

Add discount condition factory for integration tests. 

**Testing**
Unit tests
- Invalidate discount when no customer id is provided
- Invalidate discount when no customer has no groups
- Return true if groups exist on customer (assuming the group passes the check in repository layer)
- Calculate fixed discount applied on total amount
- Calculate fixed discount applied on an item basis
- Calculate percentage discount
- Return full line item price in case discount exceeds this
- Return early if item is not discountable

_Integration tests_ (added to existing test suite)
- Create discount with conditions using factory
- Create discount with conditions
- Create discount with conditions and update same conditions in a subsequent call
- Fails to create discount when multiple condition types are provided in payload
- Fails to update discount when multiple condition types are provided in payload
- Fails to add condition on rule with existing comb. of type and operator

### CODE REVIEW

File: packages/medusa/src/subscribers/order.js [hunk 1]
```diff
@@ -29,20 +29,20 @@ class OrderSubscriber {
...
 
-  handleOrderPlaced = async data => {
+  handleOrderPlaced = async (data) => {
     const order = await this.orderService_.retrieve(data.id, {
       select: ["subtotal"],
       relations: [
         "discounts",
         "discounts.rule",
-        "discounts.rule.valid_for",
+        // TODO: Add conditions relation
```
[the remark this answers is not in the document]
Comment (olivermrbl) [reply packages/medusa/src/subscribers/order.js:38]: Well spotted. I'll fix 👍

**Review by srindom — COMMENTED** [review]
Should we add tests that verify that conditions work i.e.:

- Add discount code as authenticated user that has access
- Fail when authenticated user cannot use code
- Verify that calculations with product conditions are correct

File: packages/medusa/src/api/routes/admin/discounts/create-dynamic-code.ts [hunk 2]
```diff
@@ -55,6 +60,11 @@ export class AdminPostDiscountsDiscountDynamicCodesReq {
   @IsNotEmpty()
   code: string
 
+  @IsDate()
+  @IsOptional()
+  @Type(() => Date)
+  ends_at?: Date
+
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/discounts/create-dynamic-code.ts:67]: I think this should be controlled based on the `valid_duration` or  `ends_at` fields on the "parent" discount. I.e. not configurable on creation

File: packages/medusa/src/migrations/1646324713514-discount_conditions.ts [hunk 3]
```diff
@@ -0,0 +1,269 @@
+import { MigrationInterface, QueryRunner } from "typeorm"
+
+export class discountConditions1646324713514 implements MigrationInterface {
+  name = "discountConditions1646324713514"
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `CREATE TYPE "public"."discount_condition_type_enum" AS ENUM('products', 'product_types', 'product_collections', 'product_tags', 'customer_groups')`
```
Comment (srindom) [inline packages/medusa/src/migrations/1646324713514-discount_conditions.ts:8]: rm "public"

Comment (srindom) [reply packages/medusa/src/migrations/1646324713514-discount_conditions.ts:8]: general for all below

**Review by srindom — COMMENTED** [review]
amazing work 💜

File: packages/medusa/src/api/routes/admin/discounts/create-dynamic-code.ts (hunk 2 above)
Comment (olivermrbl) [reply packages/medusa/src/api/routes/admin/discounts/create-dynamic-code.ts:67]: This must have been included by a mistake. I'll revert back 👍

**Comment by olivermrbl** [comment]
> * Add discount code as authenticated user that has access
> * Fail when authenticated user cannot use code
> * Verify that calculations with product conditions are correct

Should we coordinate on these tests @zakariaelas. I'll include the `canApplyForCustomer` check in `applyDiscount`, and add tests to verify cases related to customers. Were you planning on adding tests for the product discounts as part of the LineItemAdjustment PR?

**Comment by zakariaelas** [comment]
sure @olivermrbl! I believe the relevant tests for the line item adjustment PR are the ones that test totals and discounts calculation + the correct creation of line item adjustments when applying the discount to a cart/order. So to me, I will probably have to add a test case or two for when a discount should/should not apply for a specific customer to check whether the line item adjustments are created correctly. For me, I think having those factories that take care of "wiring up" a discount with a customer should be very helpful. What do you think?

**Comment by olivermrbl** [comment]
> For me, I think having those factories that take care of "wiring up" a discount with a customer should be very helpful. What do you think?

Makes sense @zakariaelas! So just to be sure we are not doing double work, I'll 1) add the ability to generate customer group conditions in the discount factory, and 2) add tests checking whether or not a discount with customer conditions is valid for a given customer. 

You'll then add the tests for totals and discounts calculations as well as line item adjustment creation for all cases including customers. 

Please let me know, if I am missing something 😅

**Comment by zakariaelas** [comment]
Correct @olivermrbl !

**Comment by olivermrbl** [comment]
> Should we add tests that verify that conditions work i.e.:
> 
> * Add discount code as authenticated user that has access
> * Fail when authenticated user cannot use code

@srindom Have added integration tests for:
- Create discount with customer group conditions
- Successfully apply discount when customer groups condition is defined for customer group
- Successfully apply discount when no conditions is defined for customer group
- Successfully apply discount when customer groups condition must not contain customer group
- Fails when a `not_in` condition is defined for customer group
- Fails when a `in` condition is defined but customer group is not part of it

Also added test factories for customers and customer groups.

**Review by srindom — APPROVED** [review]
LGTM!

**Pull request merged**